File: haskell/src/Day05.hs
module Day05 where

import Data.List (sort)

day05 :: IO ()
day05 = do
    input <- readFile "../data/day05.txt"
    let seats = parseInput input
    let maxSeat = maximum seats
    let mySeat = findMySeat seats
    putStrLn "\n===== Day 5 ====="
    putStrLn $ "The highest seat number is " ++ show maxSeat
    putStrLn $ "My seat number is " ++ show mySeat

parseInput :: String -> [Int]
parseInput = map (seatNr 0) . lines
    where
        seatNr n [] = n
        seatNr n (ch : chs)
            | (ch == 'F') || (ch == 'L') = seatNr (2*n) chs
            | otherwise = seatNr (2*n + 1) chs

findMySeat :: [Int] -> Int
findMySeat = findMySeat' . sort
    where
        findMySeat' (n1:n2:ns) = 
            if n2 == (n1 + 2) 
                then n1 + 1 
                else findMySeat' (n2:ns)

File: typescript/src/day05.ts
export default function day5(data: string) {
  const boardingPasses = data.split('\n');
  const seats = boardingPasses.map(getRow)
  const max = Math.max(...seats);
  const mySeat = findSingleHole(seats);
  console.log("===== Day 5 =====");
  console.log("The highest seat number is", max);
  console.log("The number of my seat is", mySeat);
}

function getRow(s: String): number {
  const binary = s.replace(/[FL]/g, '0').replace(/[BR]/g, '1');
  return parseInt(binary, 2);
}

function findSingleHole(seats: number[]): number {
  const seatSet = new Set(seats);
  for(let s = 0; s < Math.max(...seats); s++) {
    if (seatSet.has(s-1) && seatSet.has(s+1) && !seatSet.has(s)) return s;
  }
  return -1;
}

